Add property type constraints (#597)
neo4j/docs-cypher: modules/ROOT/pages/constraints/index.adoc

All queries that try to create relationships of the specified type, but without this property, will fail.
The same is true for queries that try to remove the mandatory property.

*Node property type constraints* label:enterprise-edition[]::
_This feature was introduced in Neo4j 5.9._
+
Node property type constraints ensure that a property have the required property type for all nodes with a specific label.
Queries that try to add or modify this property to nodes of the specified label, but with a different property type, will fail.

> Node property type constraints do not require all nodes to have the property (nodes without the property are not subject to this rule).

Reviewer: Consider adding something along the lines of "use these constraints in conjuction with property existence constraints if ..."

Author: We don't really have that kind of comments on previous constraints and this is the list telling us which types are available and what they do so it feels wrong to put such a note here 🤔 I'll see if I can add some note about it though.

Reviewer: It just feels incomplete to say "this thing doesn't do X" without hinting at how to do "X"

Author: It's the same for uniqueness constraints, but look at the last commit and see what you think :)

> Node property type constraints do not require all nodes to have the property (nodes without the property are not subject to this rule).

Reviewer 2: ```suggestion
Node property type constraints do not require all nodes to have the property (nodes without the property on which the constraint exists are not subject to this rule).
```

Author: also applies to uniqueness constraints

> *Relationship property type constraints* label:enterprise-edition[]::

Reviewer: (Something for a day of cleanup) This section (including the previous art) is starting to feel like a large wall of text. I feel like we could simplify it by merging the node and relationship variants into one which can then link to the examples section for details on either?

Author: The docs team is planning on restructuring the index and constraints chapters in general so I think we should wait for that and then see if anything needs further updates

_This feature was introduced in Neo4j 5.9._
+
Relationship property type constraints ensure that a property have the required property type for all relationships with a specific type.
Queries that try to add or modify this property to relationships of the specified type, but with a different property type, will fail.

> Relationship property type constraints do not require all relationships to have the property (relationships without the property are not subject to this rule).

Reviewer 2: ```suggestion
Relationship property type constraints do not require all relationships to have the property (relationships without the property on which the constraint exists are not subject to this rule).
```

Author: also applies to uniqueness constraints

[NOTE]
====
Node key constraints, relationship key constraints, node property existence constraints, relationship property existence, node property type constraints, and relationship property type constraints are not available in Neo4j Community Edition.
Databases containing one of these constraint types cannot be opened using Neo4j Community Edition.
====

Some constraint types are allowed on the same label/relationship type and property combination.

> One example is a uniqueness and an existence constraint together, though this would be the same thing as having a node or relationship key constraint.

Reviewer 2: ```suggestion
For example, it is possible to have a uniqueness and an existence constraint on the same label/relationship type and property combination, though this would be the equivalent of having a node or relationship key constraint.
```
